### ISSUE TITLE
Configure Hazelcast Server Programatically in Client/Server Deployments

### ISSUE THREAD

**Issue by gokhansari** [hazelcast/hazelcast#16527]
In client/server deployment mode, usage of programmatic configuration of **hazelcast server** might be a good option instead of declarative configuration thus programmatic configuration usage and management is more handy.

**Consider this usage recommendation**:
**1-** Introduce an interface to be used to implement for creating HZ **Config** class. 
**2-** Introduce a way to provide jar file which include related classes.
**3-** Introduce a way to select configuration class name from class path. **example**: `java -Dhazelcast.config.class-name=com.xxx.HazelcastConfiguration`
**4-** Start with this configuration instead of xml.

**Comment by mmedenjak** [comment]
Hi @gokhansari ! Thank you for the suggestion. I hope I understood your suggestion so I have one question - instead of Hazelcast providing a factory interface for creating config instances and you providing an implementation of that factory implementation, wouldn't it be easier if you simply created that `Config` instance in the first place and provided it to start Hazelcast?

Can you give us some examples of what this interface and implementation would look like?

**Comment by gokhansari** [comment]
yes you are right, it might be easier directly extend existing `Config` and provide implementation, but if you introduce this kind of interface, maybe it can be used for some other approaches, like yamlConfigurationFactory, configServerConfigurationFactory etc..

Interface:
```
public interface ConfigurationFactory {
    Config getConfiguration();
}
```
Custom Implementation:
```
public class ProgrammaticConfigurationFactory implements ConfigurationFactory{

    @Override
    public Config getConfiguration() {
        final Config config = new Config();

        config
            .getGroupConfig()
            .setName("group-name");
        
        // prepare relevant configuration
        
        return config;
    }
}
```
More;
Hazelcast can provide own classes for other configuration methods, yaml/json etc..:
```
public class YamlConfigurationFactory implements ConfigurationFactory{

    @Override
    public Config getConfiguration() {
        final Config config = new Config();
        // read yaml and prepare config file
        return config;
    }
}
```
`java -Dhazelcast.config.yaml.path=/opt/hazelcast/xxx/` 

In anyway, providing programmatically server(not-embedded) configuration flexibility in client/server mode would be useful.

**Comment by mmedenjak** [comment]
Noted, thanks. We'll consider this as a future enhancement. But for now, you can have such a factory in a wrapper application using hazelcast and prepare the `Config` instance in your code.

**Comment by gokhansari** [comment]
Thanks, just want you to aware that there might not be a wrapper application, especially while using hazelcast docker images to start a member node. In this case there is no way to prepare `Config` instance.

**Comment by blazember** [comment]
Hi @gokhansari, thank you for this proposal!

I understand that the programmatic configuration has some advantages compared to the declarative ways and that making it environment-aware is useful. While such a feature can be added to Hazelcast, it needs to be considered how much it would help the users in the scenarios you draw up. It can be argued that the cost of implementing the actual config factories and the testing is still on the user side, while it adds implementation and maintenance cost to the Hazelcast codebase as well. 

I'd also like to point out that there are some building blocks for some of the scenarios present in your examples. 

Customizing a declarative configuration programmatically can be done with the classes extending `Config` in the HZ codebase, such as `FileSystemXmlConfig`, `InMemoryYamlConfig` etc, and there is also `YamlConfigBuilder#YamlConfigBuilder(java.io.InputStream)` and its XML counterpart. Some of these classes can be used for building a custom configuration server. 

Also, for the config server scenario, the declarative configuration might be a more natural choice, since those can be put on the wire as is, while `Config` needs to be made serializable. Note that transforming a `Config` object to a declarative XML configuration can be done with `ConfigXmlGenerator`, which can be used for implementing programmatic configuration on the configuration server side and the produced config can be sent as an XML to the member wrapper application if this is a concrete case. 

How this proposal would help with the Hazelcast Docker image scenario - if there is no wrapper application - is not fully clear though. Would you mind elaborating on how the factory interface would help in this case?

**Comment by gokhansari** [comment]
> Would you mind elaborating on how the factory interface would help in this case?

factory interface was just an idea to allow developers to build their own configuration in the way they like, and I just figured out that there is **already** `ConfigBuilder` in the hazelcast code base. 

```
package com.hazelcast.config;

/**
 * Interface for all config builders.
 */

public interface ConfigBuilder {

    /**
     * Builds Config object.
     *
     * @return Built Config object
     */
    Config build();
}
```

> How this proposal would help with the Hazelcast Docker image scenario - if there is no wrapper application - is not fully clear though. 

as i mentioned at the above, If hazelcast introduce a way to pass my config class (which implements `ConfigBuilder`) on startup like this way `java -Dhazelcast.config.class-name=com.xxx.HazelcastConfiguration`, this proposal will help developer out.
